### PR TITLE
Wiegand support

### PR DESCRIPTION
**Pull request by fivegrant**

### CODE REVIEW

File: controller/api.py
Comment (hoanhan101) [inline controller/api.py:29]: if this is not gonna be used, it doesn't make sense to include it here. it's up to the controller to decide the error handling in this case.

File: controller/api.py
Comment (hoanhan101) [inline controller/api.py:50]: as mentioned above, the error handling will be put in later. i'm fine if it's just `self.wiegand.read()` for now.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py]: this file only contains the api, not the main function. this line is not necessary.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py]: make sure you follow the commenting convention. instead of `#sample comment`, this is better -> `# sample comment`.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py]: what did you choose to read the config from json in the first place? disk input/output is expensive. and you only have 3 config options anyway. it's fine to just initialize these like so:
```
self.data0 = 15
self.data1 = 13
self.facility = 32
```
also, if we're gonna hardwire it, the config can be hardcoded since it doesn't make sense to be configurable anymore.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py]: similar comment suggestion as above.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py]: you might need to do some research on how to structure these error exceptions. putting them here doesn't look so nice.

also, i'm seeing you `pass` them. why don't they return anything meaningful, like error code or message,...?

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py:21]: i think this can be optimized better. will look into this myself soon and see if i can help improve some things there.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py:45]: it would be nice if you can provide some sample reading too, in a txt file or something.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py]: as mentioned above, we don't include the main function in this package file.

**Review by hoanhan101 — CHANGES_REQUESTED** [review]
thanks for the pr! overall, good works but still need some more clean-up in order for this to get merged in. i left some comments regarding to the convention, error handling and optimization so maybe these can help you figuring out your next steps.

File: controller/api.py
Comment (fivegrant) [reply controller/api.py:29]: I put this here to show how the error handling worked. didn't expect to keep it. Then again, I should have implemented an example so heavily

File: wiegand/api.py
Comment (fivegrant) [reply wiegand/api.py]: sorry, forgot to remove that

File: wiegand/api.py
Comment (fivegrant) [reply wiegand/api.py]: I asked andrew and he said that it was fine to have a class within a class as long as it was an exception

File: wiegand/api.py
Comment (fivegrant) [reply wiegand/api.py:21]: This can be optimized once the hardware is set up. We should worry about merging this right now and experiment with optimizing later

File: wiegand/api.py
Comment (fivegrant) [reply wiegand/api.py:45]: I can do that later once the hardware is set up.

File: wiegand/api.py
Comment (fivegrant) [reply wiegand/api.py]: What main question?

Comment (fivegrant) [reply wiegand/api.py]: got rid of it

File: wiegand/api.py
Comment (hoanhan101) [reply wiegand/api.py]: i agree. it's fine to leave it here but it's just not a good practice.

File: wiegand/api.py
Comment (hoanhan101) [reply wiegand/api.py:21]: at the moment, this repo is about simulation. i wouldn't worry about setting up the hardware and get this merge right now. this code still seems messy to me. i wouldn't merge this if it's not cleaned up yet.

File: wiegand/api.py
Comment (hoanhan101) [reply wiegand/api.py:45]: when the hardware is setup, it's not about the simulation anymore. what we're trying to accomplish at the moment is to have a working simulation. that said, sample data is crucial.

File: wiegand/api.py
Comment (hoanhan101) [reply wiegand/api.py]: this `if __name__ == "__main__":`

File: controller/api.py
Comment (hoanhan101) [reply controller/api.py:29]: if you want to show an example, what you could do instead is to create an example file and put it there. you don't want to mix it with your core api.

File: wiegand/api.py
Comment (fivegrant) [reply wiegand/api.py]: should i make an exceptions.py instead?

Comment (hoanhan101) [reply wiegand/api.py]: yes, that's a good idea. i would suggest to look for how other people do it. this is a nice [example](https://github.com/kennethreitz/requests/blob/master/requests/exceptions.py). note how they structure their project too.

File: wiegand/exceptions.py
Comment (hoanhan101) [inline wiegand/exceptions.py:7]: when does it happen and how often does it happen? is it the hardware failure or software failure? if it's the software failure, can we handle it internally or correct it in some way instead of returning an error? same question for the facility error above.

File: wiegand/api.py
Comment (hoanhan101) [inline wiegand/api.py:33]: are there incoming bits after 26 bits here or are you only looking at 26 bits at a time? from what i understand from this line of code, once `capture` has 26 bits, it will look into the elif condition and do something. but the comment is saying that `there are no incoming bits` so i am not sure if that statement is true.

**Pull request closed without merging**
